site/app/models/gradeable/Component.php

```diff
@@ -0,0 +1,230 @@
+    protected $page = -1;
+
+    /** @property @var Mark[] All possible common marks that can be assigned to this component */
+    protected $marks = array();
+
+
+    public function __construct(Core $core, Gradeable $gradeable, $details, array $marks)
+    {
+        parent::__construct($core);
+
+        $this->setGradeable($gradeable);
+        $this->setMarks($marks);
+        $this->setId($details['id']);
+        $this->setTitle($details['title']);
+        $this->setTaComment($details['ta_comment']);
+        $this->setStudentComment($details['student_comment']);
+        $this->setPoints($details['lower_clamp'], $details['default'], $details['max_value'], $details['upper_clamp']);
+        $this->setText($details['text']);
+        $this->setPeer($details['peer']);
+        $this->setOrder($details['order']);
+        $this->setPage($details['page']);
+    }
+
+    public function getGradeable()
+    {
+        return $this->gradeable;
+    }
+
+    /* Overridden setters with validation */
+    private function setGradeable(Gradeable $gradeable)
+    {
+        if($gradeable === null) {
+            throw new \InvalidArgumentException('Gradeable Cannot be null!');
+        }
+        $this->gradeable = $gradeable;
+    }
+
+    private function validatePoints(&$lower_clamp, &$default, &$max_value, &$upper_clamp)
+    {
+        $errors = array();
+        if (!is_numeric($lower_clamp)) {
+            $lower_clamp = null;
+            $errors['lower_clamp'] = 'Value must be a number!';
+        } else {
+            $lower_clamp = floatval($lower_clamp);
+        }
+        if (!is_numeric($default)) {
+            $default = null;
+            $errors['default'] = 'Value must be a number!';
+        } else {
+            $default = floatval($default);
+        }
+        if (!is_numeric($max_value)) {
+            $max_value = null;
+            $errors['max_value'] = 'Value must be a number!';
+        } else {
+            $max_value = floatval($max_value);
+        }
+        if (!is_numeric($upper_clamp)) {
+            $upper_clamp = null;
+            $errors['upper_clamp'] = 'Value must be a number!';
+        } else {
+            $upper_clamp = floatval($upper_clamp);
+        }
+
+        if (!($lower_clamp === null || $default === null) && $lower_clamp > $default) {
+            $errors['lower_clamp'] = 'Lower clamp can\'t be more than default!';
+        }
+        if (!($default === null || $max_value === null) && $default > $max_value) {
+            $errors['max_value'] = 'Max value can\'t be less than default!';
+        }
+        if (!($max_value === null || $upper_clamp === null) && $max_value > $upper_clamp) {
+            $errors['max_value'] = 'Max value can\'t be more than upper clamp!';
+        }
+
+        if (count($errors) === 0)
+            return null;
+        return $errors;
+    }
+    public function setPoints($lower_clamp, $default, $max_value, $upper_clamp)
```

Comment:
Would be nice to have docs about what values are valid for this without having to look at the error messages